src/rust-crypto/backup.ts

```diff
@@ -342,20 +350,37 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
 
                 if (!request || this.stopped || !this.activeBackupVersion) {
                     logger.log(`Backup: Ending loop for version ${this.activeBackupVersion}.`);
+                    if (!request) {
+                        // nothing more to upload
+                        this.emit(CryptoEvent.KeyBackupSessionsRemaining, 0);
+                    }
                     return;
                 }
 
-                try {
-                    await this.outgoingRequestProcessor.makeOutgoingRequest(request);
-                    numFailures = 0;
-                    if (this.stopped) break;
+                // Keys count performance (`olmMachine.roomKeyCounts()`) can be pretty bad on some configurations (big database in FF).
+                // We detected on some M1 mac that when the object store reach a threshold, the count performances stops growing in O(n) and
+                // suddenly become very slow (40s, 60s or more). Even on other configurations, the count can take several seconds.
+                // This will block other operations on the database, like sending messages
+                // This is a workaround to avoid calling `olmMachine.roomKeyCounts()` too often, and only when necessary.
+                // We don't call it on the first loop because there could be only a few keys to upload, and we don't want to wait for the count.
+                if (!isFirstIteration && remainingToUploadCount === null) {
                     try {
                         const keyCount = await this.olmMachine.roomKeyCounts();
-                        const remaining = keyCount.total - keyCount.backedUp;
-                        this.emit(CryptoEvent.KeyBackupSessionsRemaining, remaining);
+                        remainingToUploadCount = keyCount.total - keyCount.backedUp;
                     } catch (err) {
                         logger.error("Backup: Failed to get key counts from rust crypto-sdk", err);
                     }
+                }
+
+                try {
+                    await this.outgoingRequestProcessor.makeOutgoingRequest(request);
+                    numFailures = 0;
+                    if (this.stopped) break;
+                    if (remainingToUploadCount !== null) {
+                        const keysCountInBatch = this.keysCountInBatch(request);
+                        remainingToUploadCount = Math.max(remainingToUploadCount - keysCountInBatch, 0);
```

Comment:
while the backup loop runs, new megolm sessions could arrive (especially if you're backing up 400K sessions or so), so presumably memoizing the count like this could mean remainingToUploadCount gets stuck at 0 for quite a while at the end (while it backs up the additional keys that arrived).  Worst case, the user might think that the progress has got stuck at zero and cancel it somehow.  I wonder if it's worth spelling this edge case out at least as a comment, that the KeyBackupSessionsRemaining event may now lie and keep emitting 0 for a bit at the end of the backup loop?

Reply:
Yes, worth explaining. Added comment here https://github.com/matrix-org/matrix-js-sdk/pull/4015/commits/8dc1ff717d10d31b63b9af7a85cc3e0d687a0252